x-pack/plugins/fleet/server/services/preconfiguration/outputs.ts

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { ElasticsearchClient, SavedObjectsClientContract } from 'src/core/server';
+import { isEqual } from 'lodash';
+import { safeDump } from 'js-yaml';
+
+import type { PreconfiguredOutput, Output } from '../../../common';
+import { normalizeHostsForAgents } from '../../../common';
+import { outputService } from '../output';
+import { agentPolicyService } from '../agent_policy';
+
+import { appContextService } from '../app_context';
+
+export async function ensurePreconfiguredOutputs(
+  soClient: SavedObjectsClientContract,
+  esClient: ElasticsearchClient,
+  outputs: PreconfiguredOutput[]
+) {
+  await createOrUpdatePreconfiguredOutputs(soClient, esClient, outputs);
+  await cleanPreconfiguredOutputs(soClient, outputs);
+}
+
+export async function createOrUpdatePreconfiguredOutputs(
+  soClient: SavedObjectsClientContract,
+  esClient: ElasticsearchClient,
+  outputs: PreconfiguredOutput[]
+) {
+  const logger = appContextService.getLogger();
+
+  if (outputs.length === 0) {
+    return;
+  }
+
+  const existingOutputs = await outputService.bulkGet(
+    soClient,
+    outputs.map(({ id }) => id),
+    { ignoreNotFound: true }
+  );
+
+  await Promise.all(
+    outputs.map(async (output) => {
+      const existingOutput = existingOutputs.find((o) => o.id === output.id);
+
+      const { id, config, ...outputData } = output;
+
+      const configYaml = config ? safeDump(config) : undefined;
+
+      const data = {
+        ...outputData,
+        config_yaml: configYaml,
+        is_preconfigured: true,
+      };
+
+      if (!data.hosts || data.hosts.length === 0) {
+        data.hosts = outputService.getDefaultESHosts();
+      }
+
+      const isCreate = !existingOutput;
+      const isUpdateWithNewData =
+        existingOutput && isPreconfiguredOutputDifferentFromCurrent(existingOutput, data);
+
+      if (isCreate) {
+        logger.debug(`Creating output ${output.id}`);
+        await outputService.create(soClient, data, { id, fromPreconfiguration: true });
+      } else if (isUpdateWithNewData) {
+        logger.debug(`Updating output ${output.id}`);
+        await outputService.update(soClient, id, data, { fromPreconfiguration: true });
+        // Bump revision of all policies using that output
+        if (outputData.is_default || outputData.is_default_monitoring) {
+          await agentPolicyService.bumpAllAgentPolicies(soClient, esClient);
+        } else {
+          await agentPolicyService.bumpAllAgentPoliciesForOutput(soClient, esClient, id);
+        }
+      }
+    })
+  );
+}
+
+export async function cleanPreconfiguredOutputs(
```

Comment:
The only change in this file is here the rest is just moving this to his own module